crates/math/src/arith_expr.rs

```diff
@@ -924,8 +677,81 @@ impl<F: Field> From<SparseLinearNormalForm<F>> for LinearNormalForm<F> {
 	}
 }
 
+struct StepNode<'a, F: Field> {
+	index: usize,
+	steps: &'a [ArithCircuitStep<F>],
+}
+
+impl<F: Field> StepNode<'_, F> {
+	fn prev_step(&self, step: usize) -> Self {
+		StepNode {
+			index: step,
+			steps: self.steps,
+		}
+	}
+}
+
+impl<F: Field> PartialEq for StepNode<'_, F> {
+	#[allow(clippy::suspicious_operation_groupings)] // false positive
+	fn eq(&self, other: &Self) -> bool {
+		match (&self.steps[self.index], &other.steps[other.index]) {
+			(ArithCircuitStep::Const(left), ArithCircuitStep::Const(right)) => left == right,
+			(ArithCircuitStep::Var(left), ArithCircuitStep::Var(right)) => left == right,
+			(
+				ArithCircuitStep::Add(left, right),
+				ArithCircuitStep::Add(other_left, other_right),
+			)
+			| (
+				ArithCircuitStep::Mul(left, right),
+				ArithCircuitStep::Mul(other_left, other_right),
+			) => {
+				self.prev_step(*left) == other.prev_step(*other_left)
+					&& self.prev_step(*right) == other.prev_step(*other_right)
+			}
+			(ArithCircuitStep::Pow(base, exp), ArithCircuitStep::Pow(other_base, other_exp)) => {
+				self.prev_step(*base) == other.prev_step(*other_base) && exp == other_exp
+			}
+			_ => false,
+		}
+	}
+}
+
+impl<F: Field> Eq for StepNode<'_, F> {}
+
+impl<F: Field> Hash for StepNode<'_, F> {
```

Comment:
Why do we need this?

Reply:
This is a helper struct to compare the subtrees by their values and store in hashed collections. This particular is needed to find the repeating subexpressions in the `deduplicate_steps`